### PR TITLE
REF: dont consolidate in BlockManager.equals

### PR DESCRIPTION
**Pull request by jbrockmendel**
Avoid side-effects.

Side-note: with EA.equals recently added to the interface, should we update array_equivalent to use it?  ATM array_equivalent starts by casting both inputs to ndarray.

### CODE REVIEW

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -1409,29 +1411,37 @@ def take(self, indexer, axis: int = 1, verify: bool = True, convert: bool = True
...
+                return array_equivalent(blk.values, rblk.values)
+
+        for i in range(len(self.items)):
+            left = self.iget_values(i)
+            right = other.iget_values(i)
+            if not is_dtype_equal(left.dtype, right.dtype):
+                return False
+            elif isinstance(left, ExtensionArray):
+                if not left.equals(right):
+                    return False
```
Comment (jorisvandenbossche) [inline pandas/core/internals/managers.py:1440]: this can be `return left.equals(right)` ?

And same for `array_equivalent` below?

**Review by jorisvandenbossche — COMMENTED** [review]
With this change, we no longer use the Block.equals methods?

**Comment by jorisvandenbossche** [comment]
> Side-note: with EA.equals recently added to the interface, should we update array_equivalent to use it?

Probably, yes

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:1440]: yes, will update

**Comment by jbrockmendel** [comment]
> With this change, we no longer use the Block.equals methods?

I think this is correct, will double-check and update.

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:1440]: no, wait, not here.  this is an "all" check, so we only return on False.

File: pandas/core/internals/managers.py [hunk 2]
```diff
@@ -1409,29 +1411,37 @@ def take(self, indexer, axis: int = 1, verify: bool = True, convert: bool = True
...
-        # the iteration order
-        def canonicalize(block):
-            return (block.mgr_locs.as_array.tolist(), block.dtype.name)
-
-        self_blocks = sorted(self.blocks, key=canonicalize)
-        other_blocks = sorted(other.blocks, key=canonicalize)
-        return all(
-            block.equals(oblock) for block, oblock in zip(self_blocks, other_blocks)
-        )
+        if self.ndim == 1:
```
Comment (jreback) [inline pandas/core/internals/managers.py:1421]: can you give some comments here (or at least a comment about the ordering of the checks)

File: pandas/core/internals/managers.py [hunk 3]
```diff
@@ -1409,29 +1411,37 @@ def take(self, indexer, axis: int = 1, verify: bool = True, convert: bool = True
...
+            blk = self.blocks[0]
+            rblk = other.blocks[0]
+            if blk.is_extension and rblk.is_extension:
+                return blk.values.equals(rblk.values)
+            elif blk.is_extension or rblk.is_extension:
+                return False
+            else:
+                return array_equivalent(blk.values, rblk.values)
+
+        for i in range(len(self.items)):
```
Comment (jreback) [inline pandas/core/internals/managers.py:1433]: same

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/internals/managers.py:1442]: Ah, yes, it's inside the for loop ..

File: pandas/core/internals/managers.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:1434]: comment added

**Comment by jbrockmendel** [comment]
gentle ping

File: pandas/tests/internals/test_internals.py [hunk 4]
```diff
@@ -389,7 +389,7 @@ def test_copy(self, mgr):
 
             # copy assertion we either have a None for a base or in case of
             # some blocks it is an array (e.g. datetimetz), but was copied
-            assert cp_blk.equals(blk)
+            tm.assert_equal(cp_blk.values, blk.values)
```
Comment (jreback) [inline pandas/tests/internals/test_internals.py:392]: should equals also work here?

Comment (jbrockmendel) [reply pandas/tests/internals/test_internals.py:392]: No, Block.equals is removed

**Review by jreback — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by TomAugspurger** [comment]
This had a large performance regression in https://pandas.pydata.org/speed/pandas/index.html#frame_methods.Equals.time_frame_float_equal?commits=77f4a4d673e2d2b3f7396a94818771d332bd913c. Is that expected?

**Comment by jorisvandenbossche** [comment]
Some slowdown can probably expected for this case since it went from block-wise to column-wise (while having a consolidated block). 

But I think a large part of the slowdown is due to `array_equivalent` being much slower for certain dtypes right now. For example, the FloatBlock had a custom `equals` method not using array_equivalent but:

```    
def equals(self, other) -> bool:
        if self.dtype != other.dtype or self.shape != other.shape:
            return False
        left, right = self.values, other.values
        return ((left == right) | (np.isnan(left) & np.isnan(right))).all()
```

and this is quite a bit slower:

```
In [5]: a = np.random.randn(N)      

In [6]: %timeit ((a == a) | (np.isnan(a) & np.isnan(a))).all()    
5.28 µs ± 366 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [8]: %timeit pd.core.dtypes.missing.array_equivalent(a, a)    
57.7 µs ± 2.2 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

**Comment by jorisvandenbossche** [comment]
I suppose that a similar "fastpath" could be added to `array_equivalent`, though?

Also for DatetimeTZBlock, there was an override of `equals` for performance:

```
    def equals(self, other) -> bool:
        # override for significant performance improvement
        if self.dtype != other.dtype or self.shape != other.shape:
            return False
        return (self.values.view("i8") == other.values.view("i8")).all()
```

**Comment by jbrockmendel** [comment]
I think the thing to do here is optimize array_equivalent.

**Comment by jbrockmendel** [comment]
see also #32339

**Pull request merged**